### ISSUE TITLE
Performance SDK has hidden method log spam

### ISSUE THREAD

**Issue by quibi-jlk** [firebase/firebase-android-sdk#1128]
### [READ] Step 1: Are you in the right place?

Issues filed here should be about bugs in __the code in this repository__. 
If you have a general question, need help debugging, or fall into some
other category use one of these other channels:

  * For general technical questions, post a question on [StackOverflow](http://stackoverflow.com/)
    with the firebase tag.
  * For general Firebase discussion, use the [firebase-talk](https://groups.google.com/forum/#!forum/firebase-talk)
    google group.
  * For help troubleshooting your application that does not fall under one
    of the above categories, reach out to the personalized
    [Firebase support channel](https://firebase.google.com/support/).

### [REQUIRED] Step 2: Describe your environment

  * Android Studio version: __3.5.3___
  * Firebase Component: performance: 19.0.2, core: 17.2.1

#### Steps to reproduce:

Any performance action (app startup, moving between screens, etc.) causes the following log spam on Android 9.0 or above:

```
W/com.foo.bar: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
W/com.foo.bar: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
W/com.foo.bar: Accessing hidden method Lsun/misc/Unsafe;->putInt(Ljava/lang/Object;JI)V (greylist, linking, allowed)
W/com.foo.bar: Accessing hidden method Lsun/misc/Unsafe;->getLong(Ljava/lang/Object;J)J (greylist,core-platform-api, linking, allowed)
W/com.foo.bar: Accessing hidden method Lsun/misc/Unsafe;->putLong(Ljava/lang/Object;JJ)V (greylist, linking, allowed)
W/com.foo.bar: Accessing hidden method Lsun/misc/Unsafe;->putInt(Ljava/lang/Object;JI)V (greylist, linking, allowed)
```

This happens very frequently and pollutes our logcat. It seems like either firebase or one of its dependencies is accessing hidden methods. See this [documentation for details](https://developer.android.com/distribute/best-practices/develop/restrictions-non-sdk-interfaces). This also causes problems in strict mode.

#### Relevant Code:

Only thing needed is to import firebase performance as a dependency.

**Comment by quibi-jlk** [comment]
One place that looks suspicious is the use of ConcurrentLinkedQueue in TaskListenerImpl. That does seem to use Unsafe calls when adding new objects to the queue. In fact, many of the java concurrent data structures seem to do things like that.

**Comment by zijianjoy** [comment]
Hello @quibi-jlk, you no longer need to specify `com.google.firebase:firebase-core` in your build.gradle file (reference: https://firebase.google.com/docs/android/setup#available-libraries). And I have tried with the specified dependency versions and StrictMode with `detectNonSdkApiUsage()`, but I cannot reproduce the logs in the description. Here are something you can try to resolve the log spam:

1. Remove `com.google.firebase:firebase-core` from build.gradle file.
2. Upgrade Performance Monitoring to the latest version `com.google.firebase:firebase-perf:19.0.5`
3. If none of the options above resolve your issue, please provide the way you configure StrictMode so we can debug more.

**Comment by quibi-jlk** [comment]
Thanks for the quick response, @zijianjoy! I'm running into some issues wrangling dependencies to remove core and upgrade perf. Specifically, We end up with a version conflict for iid in this flow...

```
Dependency failing: com.google.firebase:firebase-messaging:20.0.1 -> com.google.firebase:firebase-iid@[20.0.1], but fire
  base-iid version was 20.0.2.
```

It seems that firebase messaging and config depend on iid 20.0.1, but firebase analytics 17.2.2 has a transitive dep on iid 20.0.2. Here's the full dependency graph causing issues:

```
Dependency resolved to an incompatible version: Dependency(fromArtifactVersion=ArtifactVersion(groupId=com.google.firebase, artifactId=firebase-messaging, version=20.0.1), toArtifact=Artifact(groupId=com.google.firebase, artifactId=firebase-iid), toArtifactVersionString=[20.0.1])
Dependency Resolution Help: Displaying all currently known paths to any version of the dependency: Artifact(groupId=com.google.firebase, artifactId=firebase-iid)
NOTE: com.google.android.gms translated to c.g.a.g for brevity. Same for com.google.firebase -> c.g.f
-- :app-debugCompileClasspath-app:0.0.0 task/module dep -> c.g.f:firebase-iid@20.0.1
-- :app-debugCompileClasspath-app:0.0.0 task/module dep -> c.g.a.g:play-services-measurement-api@{strictly 17.2.2}
---- c.g.a.g:play-services-measurement-api:17.2.2 library depends -> c.g.f:firebase-iid@20.0.2
-- :app-debugCompileClasspath-app:0.0.0 task/module dep -> c.g.f:firebase-analytics@17.2.2
---- c.g.f:firebase-analytics:17.2.2 library depends -> c.g.a.g:play-services-measurement-api@17.2.2
------ c.g.a.g:play-services-measurement-api:17.2.2 library depends -> c.g.f:firebase-iid@20.0.2
-- :app-debugCompileClasspath-app:0.0.0 task/module dep -> c.g.f:firebase-analytics@{strictly 17.2.2}
---- c.g.f:firebase-analytics:17.2.2 library depends -> c.g.a.g:play-services-measurement-api@17.2.2
------ c.g.a.g:play-services-measurement-api:17.2.2 library depends -> c.g.f:firebase-iid@20.0.2
-- :app-debugCompileClasspath-app:0.0.0 task/module dep -> c.g.a.g:play-services-measurement-api@16.0.4
---- c.g.a.g:play-services-measurement-api:17.2.2 library depends -> c.g.f:firebase-iid@20.0.2
-- :app-debugCompileClasspath-app:0.0.0 task/module dep -> c.g.f:firebase-perf@19.0.5
---- c.g.f:firebase-perf:19.0.5 library depends -> c.g.f:firebase-iid@20.0.1
-- :app-debugCompileClasspath-app:0.0.0 task/module dep -> c.g.f:firebase-perf@{strictly 19.0.5}
---- c.g.f:firebase-perf:19.0.5 library depends -> c.g.f:firebase-iid@20.0.1
-- :app-debugCompileClasspath-app:0.0.0 task/module dep -> c.g.f:firebase-messaging@20.0.1
---- c.g.f:firebase-messaging:20.0.1 library depends -> c.g.f:firebase-iid@[20.0.1]
-- :app-debugCompileClasspath-app:0.0.0 task/module dep -> c.g.f:firebase-messaging@{strictly 20.0.1}
---- c.g.f:firebase-messaging:20.0.1 library depends -> c.g.f:firebase-iid@[20.0.1]
-- :app-debugCompileClasspath-app:0.0.0 task/module dep -> c.g.f:firebase-iid@{strictly 20.0.2}
-- :app-debugCompileClasspath-app:0.0.0 task/module dep -> c.g.f:firebase-config@{strictly 19.0.4}
---- c.g.f:firebase-config:19.0.4 library depends -> c.g.f:firebase-iid@20.0.1
-- :app-debugCompileClasspath-app:0.0.0 task/module dep -> c.g.f:firebase-perf@19.0.5
---- c.g.f:firebase-perf:19.0.5 library depends -> c.g.f:firebase-config@19.0.4
------ c.g.f:firebase-config:19.0.4 library depends -> c.g.f:firebase-iid@20.0.1
-- :app-debugCompileClasspath-app:0.0.0 task/module dep -> c.g.f:firebase-perf@{strictly 19.0.5}
---- c.g.f:firebase-perf:19.0.5 library depends -> c.g.f:firebase-config@19.0.4
------ c.g.f:firebase-config:19.0.4 library depends -> c.g.f:firebase-iid@20.0.1
```

**Comment by zijianjoy** [comment]
Hello @quibi-jlk, yes it is because the versions you have provided for two Firebase products (`firebase-messaging` and `firebase-analytics`) depend on different versions of `firebase-iid`. The correct combination is below:

```
com.google.firebase:firebase-analytics:17.2.2
com.google.firebase:firebase-messaging:20.1.0
``` 

You also need to verify that your app is working as expected after you have upgraded the `firebase-messaging` dependency, but I don't expect a huge difference, based on the release note here: https://firebase.google.com/support/release-notes/android#messaging_v20-1-0.

FYI, you can take a look at Firebase Bill of Materials which need only a single version number for all Firebase products: https://firebase.google.com/support/release-notes/android#bom_v24-5-0. Please note that Bill of Materials is still experimental (https://firebase.google.com/docs/android/setup#firebase-bom).

I found this great article that explains what is Bill of Materials: https://dev.to/rosariopfernandes/exploring-firebase-bill-of-materials-bom-1hij, hope it helps!

**Comment by quibi-jlk** [comment]
Thanks for the tips! I managed to work around the dependency issues by removing some deps on 3p libraries that pull in old versions of firebase core as well. Unfortunately, I ended up with the same warning logs. Our only direct firebase deps at this point are:

```
    implementation 'com.google.firebase:firebase-analytics:17.2.2'
    implementation 'com.google.firebase:firebase-perf:19.0.5'
    implementation 'com.google.firebase:firebase-iid:20.0.2'
    implementation 'com.google.firebase:firebase-messaging:20.1.0'
```

I don't need to set strictmode at all to get these warning logs, though. If I remove the firebase deps, these warnings disappear.

**Comment by zijianjoy** [comment]
Thanks for sending over the Firebase dependencies and removing 3p libraries to help debugging! I have tried with your combination but cannot reproduce the error. (I used `StrictMode.enableDefaults();` in terms of StrictMode.)

I have some ideas on the issue, based on the behavior you have described. Firebase Performance collects Foreground, Background and Screen traces automatically, so when you switch the app to background or switch screen, it will mark the end of a trace and try to send the trace to Firebase. 

To help me debug further, would you like to enable logcat logging for Firebase Performance and send over the logs? Instruction: https://firebase.google.com/docs/perf-mon/get-started-android?authuser=0#step_4_optional_view_log_messages_for_performance_events

**Comment by quibi-jlk** [comment]
Thanks again, @zijianjoy! Here's are two examples of some debug logs around occurrences of the warning logs:

```
D/FirebasePerformance: Retrieving Max Duration (in minutes) of single Session configuration value.
D/FirebasePerformance: Retrieving master flag for Firebase Performance SDK enabled configuration value.
W/FirebaseRemoteConfig: No value of type 'FirebaseRemoteConfigValue' exists for parameter key 'fpr_session_max_duration_min'.
D/FirebasePerformance: Fetched value: 'true' for key: 'fpr_enabled' from Firebase Remote Config.
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzaq' is: '240'.
W/bi.qlient.debu: Accessing hidden method Lsun/misc/Unsafe;->getLong(Ljava/lang/Object;J)J (greylist,core-platform-api, linking, allowed)
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzap' is: 'true'.
D/FirebasePerformance: Retrieving Firebase Performance SDK disabled versions configuration value.
D/FirebasePerformance: Fetched value: '' for key: 'fpr_disabled_android_versions' from Firebase Remote Config.
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzam' is: ''. Resolving value as 'false'
```

```
W/bi.qlient.debu: Accessing hidden method Lsun/misc/Unsafe;->getObject(Ljava/lang/Object;J)Ljava/lang/Object; (greylist, linking, allowed)
D/FirebasePerformance: Retrieving Max Duration (in minutes) of single Session configuration value.
D/FirebasePerformance: Retrieving master flag for Firebase Performance SDK enabled configuration value.
W/FirebaseRemoteConfig: No value of type 'FirebaseRemoteConfigValue' exists for parameter key 'fpr_session_max_duration_min'.
D/FirebasePerformance: Fetched value: 'true' for key: 'fpr_enabled' from Firebase Remote Config.
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzaq' is: '240'.
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzap' is: 'true'.
D/FirebasePerformance: Retrieving Firebase Performance SDK disabled versions configuration value.
D/FirebasePerformance: Fetched value: '' for key: 'fpr_disabled_android_versions' from Firebase Remote Config.
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzam' is: ''. Resolving value as 'false'
D/FirebasePerformance: Logging NetworkRequestMetric - https://cf-public.quibi.com/episodes/858/23901_thumb_jpg_h_960_540.jpg 54180b 444ms,
D/FirebasePerformance: Retrieving master flag for Firebase Performance SDK enabled configuration value.
D/FirebasePerformance: Fetched value: 'true' for key: 'fpr_enabled' from Firebase Remote Config.
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzap' is: 'true'.
D/FirebasePerformance: Retrieving Firebase Performance SDK disabled versions configuration value.
D/FirebasePerformance: Fetched value: '' for key: 'fpr_disabled_android_versions' from Firebase Remote Config.
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzam' is: ''. Resolving value as 'false'
D/FirebasePerformance: Retrieving network request sampling rate configuration value.
D/FirebasePerformance: Fetched value: '1.0' for key: 'fpr_vc_network_request_sampling_rate' from Firebase Remote Config.
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzak' is: '1.0'.
W/bi.qlient.debu: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
W/bi.qlient.debu: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
W/bi.qlient.debu: Accessing hidden method Lsun/misc/Unsafe;->putLong(Ljava/lang/Object;JJ)V (greylist, linking, allowed)
```

**Comment by zijianjoy** [comment]
Thank you @quibi-jlk for the logs. I try to connect the relationship between this hidden API warning and Performance Monitoring SDK, but except the behavior of making the app foreground/background switch will trigger this warning, I don't find other connections to Performance Monitoring SDK yet. I have tried the following:

1. I tried to find the hidden method from the log in greylist: https://android.googlesource.com/platform/frameworks/base/+/pie-release/config/hiddenapi-light-greylist.txt, but I don't see it, it is probably the reason I cannot reproduce the issue.
2. Looked at our source code and tried to find any `getInt`, `getLong`, `putLong` action: Since we are using `SharedPreferences` to cache configuration value, these actions happened for `SharedPreferences`, but not from `java/lang/Object` which is mentioned in the log. 

Here is something we can try:
1. Can you try to start a new project with only the Firebase dependencies you are using and share with me? In this way I can exclude any external dependencies that might affect this, and try to reproduce the issue and verify it is fixed later.
2. We depend on Firebase Remote Config, so I will add the RemoteConfig label here to see whether they have encountered anything similar in the past.

**Comment by quibi-jlk** [comment]
Thanks, @zijianjoy. I'll try to get a reduced use case next week in a sample app or something. I do see some Unsafe;-> calls in the greylist here:

https://cs.android.com/android/platform/superproject/+/master:frameworks/base/config/hiddenapi-greylist-max-o.txt;l=112612

but not the functions I'm getting warnings for, which is weird. It could be device or sdk specific, perhaps. I'm on QQ1A.200105.002.

**Comment by mkotyk-godaddy** [comment]
We're also seeing this issue.  Please re-open.

`implementation 'com.google.firebase:firebase-perf:19.0.5'`
`implementation 'com.google.firebase:firebase-messaging:20.1.0'`
`implementation 'com.google.firebase:firebase-analytics:17.2.2'`

Logcat output>
`W/godaddy.polari: Accessing hidden method Lsun/misc/Unsafe;->getObject(Ljava/lang/Object;J)Ljava/lang/Object; (greylist, linking, allowed)
    Accessing hidden method Lsun/misc/Unsafe;->putObject(Ljava/lang/Object;JLjava/lang/Object;)V (greylist, linking, allowed)
W/FirebaseRemoteConfig: No value of type 'FirebaseRemoteConfigValue' exists for parameter key 'fpr_session_max_duration_min'.
W/godaddy.polari: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
    Accessing hidden method Lsun/misc/Unsafe;->putInt(Ljava/lang/Object;JI)V (greylist, linking, allowed)
    Accessing hidden method Lsun/misc/Unsafe;->getLong(Ljava/lang/Object;J)J (greylist,core-platform-api, linking, allowed)
    Accessing hidden method Lsun/misc/Unsafe;->putLong(Ljava/lang/Object;JJ)V (greylist, linking, allowed)
`

**Comment by hoppr** [comment]
Yes same issue in Logs
W/FirebaseRemoteConfig: No value of type 'FirebaseRemoteConfigValue' exists for parameter key 'fpr_session_max_duration_min'.
2020-02-26 13:57:33.995 6138-6138/com.hoppr.androidtv.foxtel.test W

**Comment by dayron9110** [comment]
Same issue in Logs:

D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzah' is: 'false'.
D/FirebasePerformance: CollectionEnabled metadata key unknown or value not found in manifest.
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzag' is: 'null'
D/FirebasePerformance: Retrieving master flag for Firebase Performance SDK enabled configuration value.
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzap' is: 'true'.
 D/FirebasePerformance: Retrieving Firebase Performance SDK disabled versions configuration value.
 D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzam' is: ''. Resolving value as 'false'

**Comment by dayron9110** [comment]
Same issue:
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzah' is: 'false'.
D/FirebasePerformance: CollectionEnabled metadata key unknown or value not found in manifest.
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzag' is: 'null'.
D/FirebasePerformance: Retrieving master flag for Firebase Performance SDK enabled configuration value.
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzap' is: 'true'.
D/FirebasePerformance: Retrieving Firebase Performance SDK disabled versions configuration value.
D/FirebasePerformance: Config resolver result for flag: 'com.google.android.gms.internal.firebase-perf.zzam' is: ''. Resolving value as 'false'

**Comment by Mezmeraid** [comment]
Same issue

> implementation 'com.google.firebase:firebase-messaging:20.1.1'
implementation 'com.google.firebase:firebase-analytics:17.2.3'
    implementation 'com.google.firebase:firebase-crashlytics:17.0.0-beta01'
    implementation 'com.google.firebase:firebase-perf:19.0.5'

> hidden method Lsun/misc/Unsafe;->putObject(Ljava/lang/Object;JLjava/lang/Object;)V (greylist, linking, allowed)
W/FirebaseRemoteConfig: No value of type 'FirebaseRemoteConfigValue' exists for parameter key 'fpr_session_max_duration_min'.
Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
    Accessing hidden method Lsun/misc/Unsafe;->putInt(Ljava/lang/Object;JI)V (greylist, linking, allowed)
    Accessing hidden method Lsun/misc/Unsafe;->getLong(Ljava/lang/Object;J)J (greylist,core-platform-api, linking, allowed)
    Accessing hidden method Lsun/misc/Unsafe;->putLong(Ljava/lang/Object;JJ)V (greylist, linking, allowed)

**Comment by zijianjoy** [comment]
There are multiple types of logging for different reasons mentioned in this thread. I would like to triage to different issues we can get better tackle them.

To @hoppr 's [comment](https://github.com/firebase/firebase-android-sdk/issues/1128#issuecomment-591255062), this was mentioned in [issue/403](https://github.com/firebase/firebase-android-sdk/issues/403). The reason of this log is because Firebase Performance depends on Firebase Remote Config, this log happens when there is no Remote Config template set (default scenario). We are working with Remote Config team to fix this so such log will not surface in default scenario.

To @dayron9110 's [comment](https://github.com/firebase/firebase-android-sdk/issues/1128#issuecomment-592059288), this is because logcat is enabled for Firebase Performance: [reference](https://firebase.google.com/docs/perf-mon/get-started-android?authuser=0#step_4_optional_view_log_messages_for_performance_events). This is used for debugging and ideally you don't need to enable logcat for Firebase Performance. So you can set `firebase_performance_logcat_enabled` as `false` in `AndroidManifest.xml` to disable this log.

To @mkotyk-godaddy and @Mezmeraid , these logs are related to the same issue originated in this thread. Thank you for the log sample we can reproduce this issue with QQ1A.200105.002. We will update this thread once we have a fix. Thank you!

**Comment by ramanpreetSinghKhinda** [comment]
Taking on from here.

Looks like there are some restrictions in place starting Android 9 (API level 28) on which non-SDK interfaces your app can use:
https://developer.android.com/distribute/best-practices/develop/restrictions-non-sdk-interfaces

A quick run of my Test App on `Android 10` yielded this:
```
2020-03-03 15:50:33.039 12878-13189/com.my.firebase.perf.testapp D/FirebasePerformance: Logging TraceMetric - My-Trace4 2ms
2020-03-03 15:50:33.040 12878-13189/com.my.firebase.perf.testapp W/se.perf.testap: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
2020-03-03 15:50:33.040 12878-13189/com.my.firebase.perf.testapp W/se.perf.testap: Accessing hidden method Lsun/misc/Unsafe;->putInt(Ljava/lang/Object;JI)V (greylist, linking, allowed)
```

I couldn't find any usage of `putInt()` in either my **Test App** or our **Source code**.

-----

Working on it to figure out how those hidden APIs are being called by `firebase-perf`.

**Comment by lexer** [comment]
Having a similar issue with logs when firebase-perf is enabled.

```
2020-03-11 07:49:55.728 11788-11788 W/FirebaseRemoteConfig: No value of type 'FirebaseRemoteConfigValue' exists for parameter key 'fpr_session_max_duration_min'.
2020-03-11 07:49:55.728 11788-11788 W/FirebaseRemoteConfig: No value of type 'FirebaseRemoteConfigValue' exists for parameter key 'fpr_enabled'.
```

**Comment by ramanpreetSinghKhinda** [comment]
Internal Tracking the Original Issue here: http://b/152073447

-----

Thanks @lexer for reaching out with your concern however this issue is different than what the original author have posted (see [previous comment](https://github.com/firebase/firebase-android-sdk/issues/1128#issuecomment-593525094) on this as well) so I would recommend creating a separate ticket. However be assure that we are tracking this internally as well http://b/150617153. The idea is for Firebase Remote Config to provide support for `containsKey()` API so that Firebase Performance can check whether a particular key is available before accessing it and thus it will avoid the logs you are observing.

**Comment by ramanpreetSinghKhinda** [comment]
Internal Tracking the Original Issue here: b/152073447

**Comment by ramanpreetSinghKhinda** [comment]
### Analysis Result
Per analysis performed using the [veridex tool](https://developer.android.com/distribute/best-practices/develop/restrictions-non-sdk-interfaces#veridex-macos), all the accessed hidden APIs are from `com.google.protobuf` library.

TL;DR:
```
49 hidden API(s) used: 24 linked against, 25 through reflection
       44 in greylist (all usages are from Protobuf)
       0 in blacklist
       0 in greylist-max-o
       5 in greylist-max-p (none is from Fireperf SDK)
```
> Note: Run the tool on a sample app that integrates "com.google.firebase:firebase-perf:19.0.5".

### Note
 - From the [docs](https://developer.android.com/distribute/best-practices/develop/restrictions-non-sdk-interfaces#list-names), Greylist are Non-SDK interfaces that you can use as long as they are not restricted for your app's [target API level](https://developer.android.com/distribute/best-practices/develop/target-sdk).

 - From the [docs](https://developer.android.com/distribute/best-practices/develop/restrictions-non-sdk-interfaces#list-names), while you can currently use some non-SDK interfaces that are part of the greylist (depending on your app's [target API level](https://developer.android.com/distribute/best-practices/develop/target-sdk)), using any non-SDK method or field always carries a high risk of breaking your app. If your app relies on non-SDK interfaces, you should begin planning a migration to SDK interfaces or other alternatives or else [request](https://developer.android.com/distribute/best-practices/develop/restrictions-non-sdk-interfaces#feature-request) a new public API from Android.

### Concern Raised
The concern raised in this ticket is about the logs messages which are generated by Android System and I couldn't find a way to disable them.

### Next Steps
Opened a [ticket](https://github.com/protocolbuffers/protobuf/issues/7337) with Protobuf. Let's wait to hear from them and then we can track this there going forward.

**Comment by ramanpreetSinghKhinda** [comment]
Update from https://github.com/protocolbuffers/protobuf/issues/7337: Issue is being tracked internally by Protobuf Team at b/154851649. The patch is planned to go into the runtime for R release.

**Comment by ramanpreetSinghKhinda** [comment]
This should be fixed in Android R release. Please reopen if the issue persists.
